semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -22,105 +20,79 @@ module Analysis.Syntax
-  mempty = Print id
-
-instance Syntax Print where
-  iff c t e = parens (str "iff" <+> c <+> str "then" <+> t <+> str "else" <+> e)
-  noop = parens (str "noop")
-
-  bool b = parens (str (if b then "true" else "false"))
-  string = parens . text
-
-  throw e = parens (str "throw" <+> e)
-
-  let_ n v b = parens (str "let" <+> name n <+> char '=' <+> v <+> str "in" <+> b (name n))
-
-str :: String -> Print
-str = Print . showString
-
-text :: Text -> Print
-text = str . unpack
-
-char :: Char -> Print
-char = Print . showChar
-
-parens :: Print -> Print
-parens p = char '(' <> p <> char ')'
-
-(<+>) :: Print -> Print -> Print
-l <+> r = l <> char ' ' <> r
-
-infixr 6 <+>
-
-name :: Name -> Print
-name = text . formatName
+data Term
+  = Var Name
+  | Noop
+  | Iff Term Term Term
+  | Bool Bool
+  | String Text
+  | Throw Term
+  | Let Name Term Term
+  | Import (NonEmpty Text)
+  deriving (Eq, Ord, Show)
 
 
 -- Abstract interpretation
 
-eval0 :: Interpret m i -> m i
+eval0 :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m, Has S.Statement sig m) => Term -> m val
 eval0 = fix eval
 
-eval :: (Interpret m i -> m i) -> (Interpret m i -> m i)
-eval eval (Interpret f) = f eval
-
-newtype Interpret m i = Interpret { interpret :: (Interpret m i -> m i) -> m i }
-
-instance (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m) => Syntax (Interpret m val) where
-  iff c t e = Interpret (\ eval -> do
+eval
+  :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m, Has S.Statement sig m)
+  => (Term -> m val)
+  -> (Term -> m val)
+eval eval = \case
+  Var n     -> lookupEnv n >>= maybe (dvar n) fetch
+  Noop      -> dunit
+  Iff c t e -> do
     c' <- eval c
-    dif c' (eval t) (eval e))
-  noop = Interpret (const dunit)
+    dif c' (eval t) (eval e)
+  Bool b    -> dbool b
+  String s  -> dstring s
+  Throw e   -> eval e >>= ddie
+  Let n v b -> do
+    v' <- eval v
+    let' n v' (eval b)
+  Import ns -> S.simport ns >> dunit
 
-  bool b = Interpret (\ _ -> dbool b)
-  string s = Interpret (\ _ -> dstring s)
 
-  throw e = Interpret (\ eval -> eval e >>= ddie)
+evalModule0 :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m) => Term -> m (Module val)
+evalModule0 i = S.runStatement mk (eval0 i) where
+  mk msgs b = pure (Module (const b) (Set.fromList (map (\ (S.Import cs) -> name (Text.intercalate (pack ".") (toList cs))) msgs)) mempty mempty)
 
-  let_ n v b = Interpret (\ eval -> do
-    v' <- eval v
-    let' n v' (eval (b (Interpret (pure (pure v'))))))
+evalModule :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m) => (Term -> S.StatementC m val) -> (Term -> m (Module val))
+evalModule f i = S.runStatement mk (eval f i) where
+  mk msgs b = pure (Module (const b) (Set.fromList (map (\ (S.Import cs) -> name (Text.intercalate (pack ".") (toList cs))) msgs)) mempty mempty)
```

Comment:
Some let-bindings here would go a long way.

Reply:
I think I'm going to 🔥 them instead, as there's no way to provide the body, unknowns, and exports independent of the domain type anyway.